tools/create_snapshot_artifact/main.py

```diff
@@ -0,0 +1,339 @@
+                             dev_name="eth1"),
+              NetIfaceConfig(host_ip="192.168.2.1",
+                             guest_ip="192.168.2.2",
+                             tap_name="tap2",
+                             dev_name="eth2"),
+              NetIfaceConfig(host_ip="192.168.3.1",
+                             guest_ip="192.168.3.2",
+                             tap_name="tap3",
+                             dev_name="eth3")]
+
+# Allow routing requests to MMDS through eth3.
+net_iface_for_mmds = net_ifaces[3]
+# Default IPv4 address to route MMDS requests.
+IPV4_ADDRESS = '169.254.169.254'
+# VM configuration file.
+VM_CONFIG_FILE = "tools/create_snapshot_artifact/complex_vm_config.json"
+# Directory for snapshot artifacts.
+# Only use major and minor versions of the kernel.
+SNAPSHOT_ARTIFACTS_ROOT_DIR = "snapshot_artifacts"
+
+
+def compile_file(file_name, dest_path, bin_name):
+    """
+    Compile source file using gcc.
+
+    The resulted executable is placed at `/dest_path/bin_name`.
+    """
+    host_tools_path = os.path.join(os.getcwd(), 'tests/host_tools')
+
+    source_file_path = os.path.join(host_tools_path, file_name)
+    bin_file_path = os.path.join(dest_path, bin_name)
+    _gcc_compile(source_file_path, bin_file_path)
+
+    return bin_file_path
+
+
+def populate_mmds(microvm, data_store):
+    """Populate MMDS contents with json data provided."""
+    # MMDS should be empty.
+    response = microvm.mmds.get()
+    assert microvm.api_session.is_status_ok(response.status_code)
+    assert response.json() == {}
+
+    # Populate MMDS with data.
+    response = microvm.mmds.put(json=data_store)
+    assert microvm.api_session.is_status_no_content(response.status_code)
+
+    # Ensure data is persistent inside the data store.
+    response = microvm.mmds.get()
+    assert microvm.api_session.is_status_ok(response.status_code)
+    assert response.json() == data_store
+
+
+def setup_vm(context):
+    """Init microVM using context provided."""
+    root_path = context.custom['session_root_path']
+    bin_cloner_path = context.custom['bin_cloner_path']
+
+    print(
+        f"Creating snapshot of microVM with kernel {context.kernel.name()}"
+        f" and disk {context.disk.name()}."
+    )
+
+    vm = init_microvm(root_path, bin_cloner_path)
+
+    # Change kernel name to match the one in the config file.
+    kernel_full_path = os.path.join(vm.path, DEST_KERNEL_NAME)
+    shutil.copyfile(context.kernel.local_path(), kernel_full_path)
+    vm.kernel_file = kernel_full_path
+
+    rootfs_full_path = os.path.join(vm.path, context.disk.name())
+    shutil.copyfile(context.disk.local_path(), rootfs_full_path)
+    vm.rootfs_file = rootfs_full_path
+
+    return vm
+
+
+def configure_network_interfaces(microvm):
+    """Create network namespace and tap device for network interfaces."""
+    # Create network namespace.
+    run_cmd(f"ip netns add {microvm.jailer.netns}")
```

Comment:
isn't this done by the Jailer class by default?

Reply:
The Jailer does this when it's set up, in the `spawn()` logic, which is too late when building a microVM from a config file. This is because the tap devices have to be created before spawning the jailer and the network namespace has to be added before creating the tap device.